### ISSUE TITLE
GoTo Definition (etc.) support

### ISSUE THREAD

**Issue by jhossbach** [charliermarsh/ruff-lsp#23]
Hey,
I love the project and would like to eventually switch to `ruff-lsp` for my daily use. To do that, I'd like to contribute more functionality to the LSP, such as GoToDefinition, hover functionality etc.
It would be cool to have some kind of roadmap for the planned features for the LSP.

**Comment by pythops** [comment]
@charliermarsh 
This is a really essential feature for ruff-lsp to be used as an lsp server :pray:

**Comment by rassie** [comment]
What would `GoToDefinition` and hover look like in context of `ruff-lsp`? It's not a generic Python LSP server after all, it's just a linter. I'm using it as a secondary server to `python-lsp-server`, which provides all the usual perks of language servers -- maybe such setup would suit you better?

**Comment by charliermarsh** [comment]
Right now, the intent is for `ruff-lsp` to be used alongside another Python LSP, to provide behaviors like GoTo Definition. `ruff-lsp` is really focused on enabling in-editor usages of Ruff -- so taking Ruff's functionality and exposing it to various clients. As I expand Ruff, `ruff-lsp` should benefit too, sometimes "automatically" (by way of introducing new checks), and sometimes explicitly (by way of introducing new actions).

So at present, the roadmap for improving the LSP is _mostly_ a subset of the roadmap for improving Ruff.

Here are a few things that are on my mind:

- Improving the autofix messages, which are far too generic (https://github.com/charliermarsh/ruff/issues/1454).
- Adding autofix for more rules, like unused variables (https://github.com/charliermarsh/ruff/issues/1460).
- Adding more editor-friendly rules, like `flake8-simplify` (https://github.com/charliermarsh/ruff/issues/998).

I would love help with any and all of those!

Beyond these items, I do want to expand Ruff into a general autoformatter. I'll write more about that soon.

**Comment by charliermarsh** [comment]
I guess the other option would be to look into leveraging another tool (like Jedi) to power some LSP capabilities and bundle them into Ruff -- @tsugumi-sys looked into that a bit here: https://github.com/charliermarsh/ruff-lsp/issues/47.

**Comment by rchl** [comment]
I'm not sure we need another jedi-based python lsp server. Jedi is the main bottleneck of servers based on it and pyright, for example, has dramatically better performance because it's not based on jedi (not even using the same language).

So unless there is plan to re-implement Jedi-like functionality in Rust, maybe let's keep rust lean and fast and focused on linting? :)

**Comment by jfcherng** [comment]
Same. Even it's implemented, I don't see the reason why I would use it while pyright (imho) outperforms jedi. Presuming my PC has Node.js always available.

**Comment by charliermarsh** [comment]
Yeah I feel similarly. If we add support for these features, it would be by way of writing a first-class autocompletion and language server in Rust, as opposed to bolting on something else.

**Comment by lostmygithubaccount** [comment]
I see this is closed as completed but per the discussion it's not something that's been done yet? and recommended workaround would be using ruff + something else to handle this?
